pkg/deploy/controller/deployment_config_controller.go

```diff
@@ -20,73 +26,185 @@ import (
-	CreateDeployment(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error)
+// NewDeploymentConfigController creates a DeploymentConfigController using API clients. The
+// controller returned will be configured to perform a sync:
+//
+//   1. Every fullSyncPeriod.
+//   2. Any time a watch event for DeploymentConfig is observed.
+//
+// Errors are delegated to util.HandleError.
+func NewDeploymentConfigController(osClient osclient.DeploymentConfigsNamespacer, kClient kclient.ReplicationControllersNamespacer,
+	codec runtime.Codec, stop <-chan struct{}, fullSyncPeriod time.Duration) *DeploymentConfigController {
+	// Make the ListWatcher for deploymentConfigs
+	deploymentConfigLW := &deployutil.ListWatcherImpl{
+		ListFunc: func() (runtime.Object, error) {
+			return osClient.DeploymentConfigs(kapi.NamespaceAll).List(labels.Everything(), labels.Everything())
+		},
+		WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+			return osClient.DeploymentConfigs(kapi.NamespaceAll).Watch(labels.Everything(), labels.Everything(), resourceVersion)
+		},
+	}
+
+	// Make the ListWatcher for deployments
+	deploymentLW := &deployutil.ListWatcherImpl{
+		ListFunc: func() (runtime.Object, error) {
+			return kClient.ReplicationControllers(kapi.NamespaceAll).List(labels.Everything())
+		},
+		WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+			return kClient.ReplicationControllers(kapi.NamespaceAll).Watch(labels.Everything(), labels.Everything(), resourceVersion)
+		},
+	}
+
+	// Initialize the stores with reflectors
+	deploymentConfigStore := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, deploymentConfigStore).RunUntil(stop)
+
+	deploymentStore := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	cache.NewReflector(deploymentLW, &kapi.ReplicationController{}, deploymentStore).RunUntil(stop)
+
+	// Create a deploymentClient impl backed by the kube client
+	deploymentClient := &deploymentConfigControllerClientImpl{
+		CreateDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+			return kClient.ReplicationControllers(namespace).Create(deployment)
+		},
+	}
+
+	// Sync every fullSyncPeriod, or every time a deploymentConfig watch event is observed.
+	// TODO: Compress events to prevent sync spamming.
+	// TODO: Extract this into something more generic.
+	sync := make(chan struct{})
+	go func() {
+		// TODO: error handling
+		configWatch, _ := osClient.DeploymentConfigs(kapi.NamespaceAll).Watch(labels.Everything(), labels.Everything(), "")
+		ticker := time.NewTicker(fullSyncPeriod)
+		for {
+			select {
+			case <-ticker.C:
+				sync <- struct{}{}
+			case event, ok := <-configWatch.ResultChan():
+				if !ok || event.Type == watch.Error {
+					glog.Errorf("Re-establishing deploymentConfig watch")
```

Comment:
Just cargo culting a minimal bit of error handling from the Reflector, which has to deal with basically every watch failure mode. I expect any code required to deal with this stuff to get moved into a reusable chunk at some point.